File: arboles/ABBIterativo.cpp
// arbol_iterativo_procedural_completo.cpp

#include <iostream>
#include <list> // Necesario para el recorrido Post-Orden Morris

/**
 * ESTRUCTURA DEL NODO
 * ¿Qué es? Define la unidad básica de nuestro árbol. Es el "ladrillo"
 * con el que construiremos toda la estructura.
 * ¿Por qué? Necesitamos una forma de almacenar un dato y, al mismo tiempo,
 * tener "caminos" o "enlaces" a otros nodos.
 */
struct Nodo {
    // ¿Qué hace? Almacena el valor numérico del nodo.
    int dato;
    // ¿Qué hacen? Son punteros. Un puntero es como una flecha que apunta a la
    // dirección de memoria de otro nodo. 'izquierdo' apunta a un nodo con un
    // valor menor, y 'derecho' a uno con un valor mayor.
    Nodo* izquierdo;
    Nodo* derecho;

    // ¿Qué es? Un constructor. Es una función especial que se ejecuta al crear un
    // nuevo Nodo.
    // ¿Por qué? Facilita la vida. Con una sola línea (ej: new Nodo(10)), crea el
    // nodo, le asigna el valor y se asegura de que sus hijos apunten a 'nada'
    // (nullptr) inicialmente.
    Nodo(int valor) : dato(valor), izquierdo(nullptr), derecho(nullptr) {}
};

/**
 * INSERCIÓN ITERATIVA
 * @param raiz La raíz del árbol. Se pasa por referencia (Nodo*&) para poder
 * modificarla directamente si el árbol está vacío.
 * @param valor El dato a insertar.
 */
void insertar(Nodo*& raiz, int valor) {
    // --- PASO 1: PREPARACIÓN - CREAR EL NUEVO NODO ---
    // ¿Qué hace? Creamos el nodo que queremos insertar en la memoria.
    // ¿Por qué? Antes de encontrar dónde ponerlo, necesitamos tener el nodo
    // listo. En este momento, es un nodo "flotante", no está conectado a nada.
    Nodo* nuevoNodo = new Nodo(valor);

    // --- PASO 2: EL CASO MÁS SIMPLE - EL ÁRBOL VACÍO ---
    // ¿Qué hace? Comprueba si el puntero de la raíz es nulo.
    // ¿Por qué? Si el árbol no tiene ningún nodo, no hay necesidad de buscar.
    // El nuevo nodo se convierte en la primera y única cosa en el árbol: la raíz.
    if (!raiz) {
        raiz = nuevoNodo;
        return;
    }

    // --- PASO 3: PREPARANDO LA BÚSQUEDA - LOS PUNTEROS GUÍA ---
    // ¿Qué hace? 'actual' se inicializa en la raíz. Será nuestro explorador.
    // ¿Por qué? Lo usaremos para descender por el árbol, moviéndose de nodo en nodo.
    Nodo* actual = raiz;
    // ¿Qué hace? 'padre' se inicializa en nulo. Siempre irá un paso por detrás de 'actual'.
    // ¿Por qué? Es la clave. Cuando 'actual' encuentre un espacio vacío (se vuelva nullptr),
    // 'padre' estará apuntando al nodo al que debemos conectar nuestro 'nuevoNodo'.
    Nodo* padre = nullptr;

    // --- PASO 4: EL DESCENSO POR EL ÁRBOL - EL BUCLE 'while' ---
    // ¿Qué hace? Recorre el árbol hasta que 'actual' se convierta en nullptr.
    // ¿Por qué? Así encontramos el final de una rama, el lugar exacto para insertar.
    while (actual != nullptr) {
        // ¿Qué hace? 'padre' se actualiza para que apunte a 'actual'.
        // ¿Por qué? ANTES de que 'actual' se mueva hacia abajo, hacemos que 'padre'
        // lo alcance. Así, 'padre' siempre recuerda el nodo que acabamos de dejar.
        padre = actual;

        // ¿Qué hace? Compara el valor a insertar con el dato del nodo 'actual'.
        // ¿Por qué? Esta es la regla del ABB. Nos dice si debemos ir a la izquierda o a la derecha.
        if (valor < actual->dato) {
            actual = actual->izquierdo;
        } else if (valor > actual->dato) {
            actual = actual->derecho;
        } else {
            // El valor ya existe. No permitimos duplicados.
            // ¿Qué hace? Elimina el 'nuevoNodo' que creamos y no vamos a usar.
            // ¿Por qué? Para evitar una fuga de memoria.
            delete nuevoNodo;
            return;
        }
    }

    // --- PASO 5: LA CONEXIÓN FINAL - ENGANCHAR EL NODO ---
    // ¿Qué hace? Compara el valor con el del 'padre' para decidir dónde conectar.
    // ¿Por qué? Ya encontramos al padre correcto. Ahora solo necesitamos conectar
    // el 'nuevoNodo' en la rama correcta para mantener el orden del ABB.
    if (valor < padre->dato) {
        padre->izquierdo = nuevoNodo;
    } else {
        padre->derecho = nuevoNodo;
    }
}

/**
 * ELIMINACIÓN ITERATIVA
 * @param raiz La raíz del árbol. Se pasa por referencia para poder modificarla.
 * @param valor El dato del nodo a eliminar.
 */
void eliminar(Nodo*& raiz, int valor) {
    if (!raiz) return;

    // --- PASO 1: BÚSQUEDA DEL NODO A ELIMINAR Y SU PADRE ---
    // ¿Qué hacen? 'padre' y 'actual' funcionan igual que en la inserción.
    // ¿Por qué? Necesitamos encontrar el nodo a borrar ('actual') y también
    // quién es su 'padre' para poder "reconectar" los punteros después.
    Nodo* padre = nullptr;
    Nodo* actual = raiz;
    while (actual != nullptr && actual->dato != valor) {
        padre = actual;
        if (valor < actual->dato) {
            actual = actual->izquierdo;
        } else {
            actual = actual->derecho;
        }
    }
    if (actual == nullptr) return; // Si no se encontró, no hay nada que hacer.

    // --- PASO 2: LÓGICA DE ELIMINACIÓN ---
    // CASO 1 (hoja) Y CASO 2 (un hijo)
    if (actual->izquierdo == nullptr || actual->derecho == nullptr) {
        // ¿Qué hace? 'hijo' apuntará al único hijo de 'actual', o a nullptr si 'actual' es una hoja.
        // ¿Por qué? Simplifica el código. Sea cual sea el caso (0 o 1 hijo), 'hijo'
        // contiene lo que debe reemplazar a 'actual'.
        Nodo* hijo = (actual->izquierdo != nullptr) ? actual->izquierdo : actual->derecho;

        // ¿Qué hace? Comprueba si el nodo a eliminar es la raíz.
        // ¿Por qué? Si no tiene padre, la raíz del árbol debe actualizarse directamente.
        if (padre == nullptr) {
            raiz = hijo;
        }
        // ¿Qué hace? Comprueba si 'actual' es un hijo izquierdo.
        // ¿Por qué? Para saber si debemos modificar el puntero 'izquierdo' o 'derecho' del padre.
        else if (actual == padre->izquierdo) {
            padre->izquierdo = hijo; // El padre "adopta" al nieto.
        } else {
            padre->derecho = hijo; // El padre "adopta" al nieto.
        }
        // ¿Qué hace? Libera la memoria del nodo que ya no está en el árbol.
        // ¿Por qué? Es crucial en C++ para evitar fugas de memoria.
        delete actual;
    }
    // CASO 3: EL NODO TIENE DOS HIJOS
    else {
        // Estrategia: Reemplazar el valor, no el nodo.
        // 1. Encontrar el sucesor (el nodo más pequeño del subárbol derecho).
        // ¿Por qué el sucesor? Porque es el único valor que garantiza ser mayor que todo
        // lo que está a la izquierda y menor que el resto de la derecha.
        Nodo* padreSucesor = actual;
        Nodo* sucesor = actual->derecho;
        while (sucesor->izquierdo != nullptr) {
            padreSucesor = sucesor;
            sucesor = sucesor->izquierdo;
        }

        // 2. Copiamos el valor del sucesor al nodo que queremos "eliminar".
        actual->dato = sucesor->dato;

        // 3. Ahora, el problema se reduce a eliminar el nodo sucesor, que es mucho más fácil
        // (siempre será un caso 1 o 2).
        if (sucesor == padreSucesor->derecho) {
            padreSucesor->derecho = sucesor->derecho;
        } else {
            padreSucesor->izquierdo = sucesor->derecho;
        }
        delete sucesor;
    }
}

/**
 * RECORRIDO IN-ORDEN CON MORRIS
 */
void recorridoInOrdenMorris(Nodo* raiz) {
    std::cout << "In-Orden (Morris):   ";
    Nodo* actual = raiz;
    while (actual != nullptr) {
        if (actual->izquierdo == nullptr) {
            // Si no hay izquierda, visitamos el nodo actual y avanzamos a la derecha.
            std::cout << actual->dato << " ";
            actual = actual->derecho;
        } else {
            // Buscamos el predecesor (el nodo más a la derecha del subárbol izquierdo).
            Nodo* predecesor = actual->izquierdo;
            while (predecesor->derecho != nullptr && predecesor->derecho != actual) {
                predecesor = predecesor->derecho;
            }

            if (predecesor->derecho == nullptr) {
                // Es la primera visita. Creamos el hilo y bajamos por la izquierda.
                predecesor->derecho = actual;
                actual = actual->izquierdo;
            } else {
                // Regresamos por el hilo. Lo rompemos, visitamos el nodo y vamos a la derecha.
                predecesor->derecho = nullptr;
                std::cout << actual->dato << " ";
                actual = actual->derecho;
            }
        }
    }
    std::cout << std::endl;
}

/**
 * RECORRIDO PRE-ORDEN CON MORRIS
 */
void recorridoPreOrdenMorris(Nodo* raiz) {
    std::cout << "Pre-Orden (Morris):  ";
    Nodo* actual = raiz;
    while (actual != nullptr) {
        if (actual->izquierdo == nullptr) {
            std::cout << actual->dato << " ";
            actual = actual->derecho;
        } else {
            Nodo* predecesor = actual->izquierdo;
            while (predecesor->derecho != nullptr && predecesor->derecho != actual) {
                predecesor = predecesor->derecho;
            }

            if (predecesor->derecho == nullptr) {
                // La diferencia clave: Visitamos el nodo al crear el hilo.
                std::cout << actual->dato << " ";
                predecesor->derecho = actual;
                actual = actual->izquierdo;
            } else {
                predecesor->derecho = nullptr;
                actual = actual->derecho;
            }
        }
    }
    std::cout << std::endl;
}

/**
 * RECORRIDO POST-ORDEN CON MORRIS
 */
void recorridoPostOrdenMorris(Nodo* raiz) {
    std::cout << "Post-Orden (Morris): ";
    std::list<int> resultado;
    Nodo* actual = raiz;
    while (actual != nullptr) {
        if (actual->derecho == nullptr) {
            resultado.push_front(actual->dato);
            actual = actual->izquierdo;
        } else {
            Nodo* sucesor = actual->derecho;
            while (sucesor->izquierdo != nullptr && sucesor->izquierdo != actual) {
                sucesor = sucesor->izquierdo;
            }
            if (sucesor->izquierdo == nullptr) {
                resultado.push_front(actual->dato);
                sucesor->izquierdo = actual;
                actual = actual->derecho;
            } else {
                sucesor->izquierdo = nullptr;
                actual = actual->izquierdo;
            }
        }
    }
    for (int dato : resultado) {
        std::cout << dato << " ";
    }
    std::cout << std::endl;
}

/**
 * LIBERACIÓN DE MEMORIA
 * @param nodo La raíz del árbol a destruir.
 */
void destruirArbol(Nodo* nodo) {
    // ¿Qué hace? Usa un recorrido Post-Orden recursivo.
    // ¿Por qué? Es la única forma segura. Debemos eliminar los hijos ANTES de
    // eliminar al padre. Si elimináramos al padre primero, perderíamos los
    // punteros a sus hijos y esa memoria nunca se liberaría (fuga de memoria).
    if (nodo) {
        destruirArbol(nodo->izquierdo);
        destruirArbol(nodo->derecho);
        delete nodo;
    }
}

int main() {
    // 1. CREACIÓN: La raíz del árbol empieza siendo un puntero a nada.
    Nodo* raiz = nullptr;

    std::cout << "--- ARBOL ITERATIVO PROCEDURAL EN C++ ---" << std::endl;
    int valores[] = {10, 5, 15, 3, 7, 12, 18};
    std::cout << "Insertando: ";
    // 2. POBLACIÓN: Se insertan los valores uno por uno en el árbol.
    for (int valor : valores) {
        std::cout << valor << " ";
        insertar(raiz, valor);
    }
    std::cout << "\n" << std::endl;

    // 3. RECORRIDOS: Se muestran los datos del árbol en diferentes órdenes.
    recorridoInOrdenMorris(raiz);
    recorridoPreOrdenMorris(raiz);
    recorridoPostOrdenMorris(raiz);

    // 4. ELIMINACIÓN: Se prueba la función de borrado.
    std::cout << "\nEliminando el 15 (nodo con dos hijos)..." << std::endl;
    eliminar(raiz, 15);
    recorridoInOrdenMorris(raiz);

    // 5. DESTRUCCIÓN: Se libera toda la memoria asignada antes de terminar.
    // ¿Por qué? Para ser un programa limpio y sin fugas de memoria.
    destruirArbol(raiz);

    return 0;
}

File: arboles/ABBIterativo.java

import java.util.LinkedList;

/**
 * ============================================================================
 *   🌲 ÁRBOL BINARIO DE BÚSQUEDA (ABB) - IMPLEMENTACIÓN ITERATIVA 🌲
 * ============================================================================
 *
 * ¿QUÉ ES ESTE ARCHIVO?
 * Este archivo contiene la implementación "mecánica" y extremadamente eficiente
 * en memoria de un ABB. En lugar de usar llamadas recursivas, utiliza bucles
 * clásicos (while) e inteligentemente dos punteros para viajar a pie por los
 * nodos del árbol.
 *
 * ¿POR QUÉ USAR UNA VERSIÓN ITERATIVA?
 * La recursividad es hermosa y el código queda muy corto (Arbol.java), pero
 * tiene un defecto mortal para árboles extremadamente desbalanceados: si ingresas
 * 100,000 números ordenados (1, 2, 3, 4...), tendrías un árbol sin hijos
 * izquierdos, literalemente una línea recta gigante.
 *
 * La recursividad haría 100,000 "pausas" o llamadas a función, consumiendo toda
 * tu Pila de Llamadas del sistema y causando un error trágico (Stack Overflow).
 * El enfoque iterativo navega el árbol usando memoria constante O(1), no importa
 * lo deforme que se vuelva. Protege la memoria a cambio de código más extenso.
 *
 * ----------------------------------------------------------------------------
 * 🚀 EJEMPLO VISUAL PASO A PASO: INSERTAR EL NÚMERO 7
 * ----------------------------------------------------------------------------
 * Usamos un "vehículo" explorador llamado `actual` y un "copiloto" llamado
 * `padre` que recuerda la última posición válida antes de caer al vacío.
 * 
 *         [10]
 *        /    \
 *      [5]    [15]
 *        \
 *        [8]
 *
 * INICIO DEL VIAJE (Bucle While buscando un nulo):
 * `actual` apunta al [10]. `padre` empieza en null.
 * 
 * PASO 1 (bucle en 10):
 *        `padre` copia la posición del `actual`. Ahora `padre` = 10.
 *         ¿7 < 10? SÍ. `actual` avanza a su izquierda (que es 5).
 *
 * PASO 2 (bucle en 5):
 *         `padre` se queda estacionado en tu última posición (`padre` = 5).
 *         ¿7 < 5? NO. ¿7 > 5? SÍ. `actual` avanza a su derecha (que es 8).
 *
 * PASO 3 (bucle en 8):
 *         `padre` toma nota (`padre` = 8).
 *         ¿7 < 8? SÍ. `actual` avanza a su izquierda...
 * 
 * ¡BUM! `actual` CAYÓ EN NULL. El bucle while termina, deteniendo el coche.
 * Sin embargo, el `padre` nos salvó, recordando que antes de caer estábamos
 * en el 8.
 *
 * EL DESENLACE:
 * Solo queda conectar el [7] nuevo adonde nos diga el copiloto `padre`.
 * Si el 7 es menor al 8, engánchalo a su izquierda. ¡Listo!
 *         [8]
 *        /
 *      [7] <-- ¡Nuevo nodo conectado perfectamente!
 * ============================================================================
 */
public class ABBIterativo {

    static class Nodo {

        int dato;
        Nodo izquierdo;
        Nodo derecho;

        public Nodo(int dato) {
            this.dato = dato;
            this.izquierdo = null;
            this.derecho = null;
        }
    }

    Nodo raiz;

    public ABBIterativo() {
        this.raiz = null;
    }

    /**
     * Inserta un valor en el árbol de forma iterativa.
     * 
     * ¿Qué se busca? Encontrar la ubicación correcta para el nuevo valor sin romper
     * la regla del ABB. Un nuevo nodo siempre se insertará como una "hoja" (un nodo sin hijos).
     * 
     * ¿Cómo se logra? Se desciende por el árbol de nodo en nodo, usando un bucle while,
     * hasta encontrar un lugar vacío (un puntero null) donde enganchar el nuevo nodo.
     */
    public void insertar(int valor) {
        Nodo nuevoNodo = new Nodo(valor);

        // --- CASO ESPECIAL: ÁRBOL VACÍO ---
        // Si no hay raíz, el nuevo nodo se convierte en la raíz y terminamos.
        if (raiz == null) {
            raiz = nuevoNodo;
            return;
        }

        // --- BÚSQUEDA DE LA POSICIÓN ---
        // Usamos dos punteros: 'actual' para explorar y 'padre' para recordar
        // el nodo al que nos conectaremos. 'padre' siempre está un paso detrás.
        Nodo actual = raiz;
        Nodo padre = null;

        // Este bucle es el corazón de la inserción. Desciende por el árbol.
        while (actual != null) {
            // Antes de mover 'actual', guardamos su posición en 'padre'.
            // Cuando 'actual' se vuelva null, 'padre' será el nodo correcto.
            padre = actual;

            if (valor < actual.dato) {
                // El valor es menor, así que debe ir en el subárbol izquierdo.
                actual = actual.izquierdo;
            } else if (valor > actual.dato) {
                // El valor es mayor, debe ir en el subárbol derecho.
                actual = actual.derecho;
            } else {
                // El valor ya existe. En un ABB simple, no se permiten duplicados.
                return;
            }
        }

        // --- CONEXIÓN DEL NUEVO NODO ---
        // Al salir del bucle, 'actual' es null, pero 'padre' apunta al
        // nodo hoja que será el padre del nuevo nodo.
        if (valor < padre.dato) {
            // Si el valor es menor que el del padre, se convierte en su hijo izquierdo.
            padre.izquierdo = nuevoNodo;
        } else {
            // Si es mayor, se convierte en su hijo derecho.
            padre.derecho = nuevoNodo;
        }
    }

    /**
     * Elimina un nodo del árbol de forma iterativa.
     *
     * ¿Qué se busca? Quitar un nodo y "reparar" la estructura para que la regla
     * del ABB se mantenga.
     *
     * ¿Por qué hay diferentes casos? La reparación depende de cuántos hijos
     * tiene el nodo a eliminar. La complejidad varía si tiene 0, 1 o 2 hijos.
     */
    public void eliminar(int valor) {
        if (raiz == null) {
            return;
        }

        // --- PASO 1: BÚSQUEDA DEL NODO A ELIMINAR Y SU PADRE ---
        Nodo padre = null;
        Nodo actual = raiz;

        // Bucle para localizar el nodo 'actual' que contiene el valor y
        // mantener una referencia a su 'padre'.
        while (actual != null && actual.dato != valor) {
            padre = actual;
            if (valor < actual.dato) {
                actual = actual.izquierdo;
            } else {
                actual = actual.derecho;
            }
        }

        // Si actual es null, el valor no estaba en el árbol.
        if (actual == null) {
            return;
        }

        // --- PASO 2: LÓGICA DE ELIMINACIÓN BASADA EN EL NÚMERO DE HIJOS ---
        // --- CASO 1 (hoja) Y CASO 2 (un hijo) ---
        // Estos dos casos se pueden manejar con la misma lógica.
        // La idea es "puentear" el nodo a eliminar. El padre adoptará al nieto.
        if (actual.izquierdo == null || actual.derecho == null) {
            Nodo hijo;

            // Determina cuál es el único hijo (si existe).
            // Si 'actual' es una hoja, 'hijo' será null.
            if (actual.izquierdo != null) {
                hijo = actual.izquierdo;
            } else {
                hijo = actual.derecho;
            }

            // Subcaso: se está eliminando la raíz.
            if (padre == null) {
                raiz = hijo;
            } // Si 'actual' es un hijo izquierdo, su padre ahora apunta a 'hijo'.
            else if (actual == padre.izquierdo) {
                /*
                 * (padre)             (padre)
                 * |                   |
                 * (actual)    --->    (hijo)
                 * |
                 * (hijo)
                 */
                padre.izquierdo = hijo;
            } // Si 'actual' es un hijo derecho, su padre ahora apunta a 'hijo'.
            else {
                padre.derecho = hijo;
            }
        } // --- CASO 3: EL NODO TIENE DOS HIJOS ---
        // Este es el caso más complejo. No podemos simplemente eliminarlo.
        // Estrategia: Reemplazar el valor del nodo a eliminar con un valor
        // adecuado de uno de sus subárboles y luego eliminar ese nodo sustituto.
        // El sustituto ideal es el "sucesor in-orden": el valor más pequeño
        // del subárbol derecho.
        else {
            // 1. Encontrar el sucesor (el nodo más a la izquierda del subárbol derecho).
            Nodo padreSucesor = actual;
            Nodo sucesor = actual.derecho;
            while (sucesor.izquierdo != null) {
                padreSucesor = sucesor;
                sucesor = sucesor.izquierdo;
            }

            // 2. Copiar el valor del sucesor al nodo que originalmente queríamos "eliminar".
            // Ahora, lógicamente, el nodo 'actual' está a salvo, pero tenemos un duplicado.
            actual.dato = sucesor.dato;

            // 3. Eliminar el nodo sucesor de su posición original.
            // Esta es una eliminación mucho más fácil, porque el sucesor, por definición,
            // tendrá como máximo un hijo (un hijo derecho), cayendo en el Caso 1 o 2.
            if (sucesor == padreSucesor.izquierdo) {
                padreSucesor.izquierdo = sucesor.derecho;
            } else {
                padreSucesor.derecho = sucesor.derecho;
            }
        }
    }

    /**
     * RECORRIDOS DE MORRIS (SIN PILA NI RECURSIÓN)
     * ¿Qué se busca? Recorrer el árbol iterativamente sin sobrecargar la
     * memoria con un Stack de llamadas.
     *
     * ¿Cómo se logra la proeza? A través de hilos (Threads mágicos).
     * Se modifica temporalmente el árbol en tiempo real creando "hilos" o uniones
     * entre el último nodo derecho inferior y la raíz. Son como cuerdas para
     * no perdernos y poder regresar sin que nadie se acuerde con la Pila de dónde
     * estábamos.
     * Al terminar de explorar esa rama izquierda, el hilo se corta para dejar el
     * árbol original intacto. Es ingeniería pura en estructuras de datos con un
     * costo exacto de memoria O(1) literal.
     */
    public void recorridoInOrdenMorris() {
        System.out.print("In-Orden (Morris):   ");
        Nodo actual = raiz;
        while (actual != null) {
            // Si no hay subárbol izquierdo, es simple: visitamos el nodo y vamos a la derecha.
            if (actual.izquierdo == null) {
                System.out.print(actual.dato + " ");
                actual = actual.derecho;
            } else {
                // Si hay subárbol izquierdo, buscamos a su predecesor in-orden.
                // El predecesor es el nodo más a la derecha del subárbol izquierdo.
                Nodo predecesor = actual.izquierdo;
                while (predecesor.derecho != null && predecesor.derecho != actual) {
                    predecesor = predecesor.derecho;
                }

                // Si el puntero derecho del predecesor es null, es nuestra primera visita.
                if (predecesor.derecho == null) {
                    // --- CREAR HILO ---
                    // Creamos un enlace temporal del predecesor de vuelta a 'actual'.
                    predecesor.derecho = actual;
                    // Ahora, descendemos por la izquierda para procesar ese subárbol.
                    actual = actual.izquierdo;
                } // Si el puntero derecho ya apunta a 'actual', significa que el hilo ya existía.
                // Esto nos indica que ya terminamos de recorrer el subárbol izquierdo.
                else {
                    // --- QUITAR HILO ---
                    predecesor.derecho = null; // Restauramos el puntero original del árbol.
                    // AHORA es el momento de visitar el nodo en In-Orden (Izquierda -> *Raíz* -> Derecha).
                    System.out.print(actual.dato + " ");
                    // Ya terminamos con la izquierda y la raíz, ahora vamos a la derecha.
                    actual = actual.derecho;
                }
            }
        }
        System.out.println();
    }

    public void recorridoPreOrdenMorris() {
        System.out.print("Pre-Orden (Morris):  ");
        Nodo actual = raiz;
        while (actual != null) {
            if (actual.izquierdo == null) {
                System.out.print(actual.dato + " ");
                actual = actual.derecho;
            } else {
                Nodo predecesor = actual.izquierdo;
                while (predecesor.derecho != null && predecesor.derecho != actual) {
                    predecesor = predecesor.derecho;
                }

                if (predecesor.derecho == null) {
                    // LA DIFERENCIA CLAVE: En Pre-Orden (*Raíz* -> Izquierda -> Derecha),
                    // visitamos el nodo la PRIMERA vez que lo encontramos, ANTES de crear
                    // el hilo y descender por la izquierda.
                    System.out.print(actual.dato + " ");
                    predecesor.derecho = actual;
                    actual = actual.izquierdo;
                } else {
                    predecesor.derecho = null;
                    actual = actual.derecho;
                }
            }
        }
        System.out.println();
    }

    public void recorridoPostOrdenMorris() {
        System.out.print("Post-Orden (Morris): ");
        // Lógica: Post-Orden directo (Izquierda -> Derecha -> Raíz) es muy complejo con Morris.
        // El truco es hacer un recorrido modificado (Raíz -> Derecha -> Izquierda)
        // y añadir cada elemento visitado al FRENTE de una lista. Esto revierte el orden.
        // El resultado de revertir (Raíz, Derecha, Izquierda) es (Izquierda, Derecha, Raíz).
        LinkedList<Integer> resultado = new LinkedList<>();
        Nodo actual = raiz;
        while (actual != null) {
            // Simétrico al In-Orden, pero trabajando con el subárbol derecho.
            if (actual.derecho == null) {
                resultado.addFirst(actual.dato);
                actual = actual.izquierdo;
            } else {
                // Buscamos el sucesor (el nodo más a la izquierda del subárbol derecho).
                Nodo sucesor = actual.derecho;
                while (sucesor.izquierdo != null && sucesor.izquierdo != actual) {
                    sucesor = sucesor.izquierdo;
                }
                if (sucesor.izquierdo == null) {
                    // Visitamos (añadimos al resultado) y creamos el hilo.
                    resultado.addFirst(actual.dato);
                    sucesor.izquierdo = actual;
                    actual = actual.derecho;
                } else {
                    // Rompemos el hilo y avanzamos.
                    sucesor.izquierdo = null;
                    actual = actual.izquierdo;
                }
            }
        }
        // Imprimimos el resultado final que ya está en el orden correcto.
        for (int dato : resultado) {
            System.out.print(dato + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ArbolBinarioBusquedaIterativo arbol = new ArbolBinarioBusquedaIterativo();
        System.out.println("--- ÁRBOL ITERATIVO ---");
        int[] valores = {10, 5, 15, 3, 7, 12, 18};
        System.out.print("Insertando: ");
        for (int valor : valores) {
            System.out.print(valor + " ");
            arbol.insertar(valor);
        }
        System.out.println("\n");

        arbol.recorridoInOrdenMorris();
        arbol.recorridoPreOrdenMorris();
        arbol.recorridoPostOrdenMorris();

        System.out.println("\nEliminando el 15 (nodo con dos hijos)...");
        arbol.eliminar(15);
        arbol.recorridoInOrdenMorris();
    }
}
